In [1]:
# import holder
from sqlalchemy import create_engine, text
import pandas as pd
import datetime as dt
import seaborn as sns

In [2]:
# bring in postgres database

database_name = 'scooters'
connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"
engine = create_engine(connection_string)

In [3]:
# header queries

scooters_sample = '''SELECT * FROM scooters LIMIT 10'''
trips_sample = '''SELECT * FROM trips limit 10'''

with engine.connect() as connection:
    scooters = pd.read_sql(text(scooters_sample), con = connection)
    trips = pd.read_sql(text(trips_sample), con = connection)

In [4]:
scooters.head()

pubdatetime   latitude  longitude        sumdid sumdtype  \
0 2019-05-01 00:01:41.247  36.136822 -86.799877  PoweredLIRL1  Powered   
1 2019-05-01 00:01:41.247  36.191252 -86.772945  PoweredXWRWC  Powered   
2 2019-05-01 00:01:41.247  36.144752 -86.806293  PoweredMEJEH  Powered   
3 2019-05-01 00:01:41.247  36.162056 -86.774688  Powered1A7TC  Powered   
4 2019-05-01 00:01:41.247  36.150973 -86.783109  Powered2TYEF  Powered   

   chargelevel sumdgroup  costpermin companyname  
0         93.0   scooter         0.0        Bird  
1         35.0   scooter         0.0        Bird  
2         90.0   scooter         0.0        Bird  
3         88.0   scooter         0.0        Bird  
4         98.0   scooter         0.0        Bird

In [5]:
trips.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance   startdate        starttime     enddate  \
0        3.0000     958.00528  2019-05-01  00:00:20.460000  2019-05-01   
1        1.7156    1371.39112  2019-05-01  00:01:50.090000  2019-05-01   
2        3.0000    2296.58800  2019-05-01  00:03:47.363333  2019-05-01   
3        3.0000    1200.78744  2019-05-01  00:04:21.386666  2019-05-01   
4        2.0000     351.04988  2019-05-01  00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0  00:02:52.346666       36.15710       -86.80360     36.15660     -86.80670   
1  00:03:33.026666       36.15797       -86.77896     36.16054     -86.77689   
2  00:07:13.596666       36.15470       -86.78180     36.15650     -86.78680   
3  00:06:59.176666       36.14940       -86.77950     36.15310     -86.77960   
4  00:06:23.150000       36.17780       -86.78660     36.17740     -86.78760   

                                           triproute               create_dt  
0  [(36.157235, -86.803612), (36.157235, -86.8036... 2019-05-02 05:30:23.780  
1  [(36.15797, -86.77896), (36.15795, -86.77873),... 2019-05-02 07:20:32.757  
2  [(36.155068, -86.782124), (36.156597, -86.78675)] 2019-05-02 05:30:24.530  
3  [(36.149741, -86.779344), (36.149741, -86.7793... 2019-05-02 05:30:24.237  
4  [(36.177699, -86.786477), (36.177711, -86.7864... 2019-05-02 05:30:24.207

In [6]:
# parse scooter data to csvs. Uncomment lines below to write file to project directory; recomment when complete to avoid running query multiple times.
# Save the dataframe to a CSV file (example: df.to_csv('Biodata.csv', index=False))

#company queries
#bird_query = '''SELECT * FROM scooters WHERE companyname ILIKE 'bird';'''
#bolt_query = '''SELECT * FROM scooters WHERE companyname ilike 'bolt';'''
#gotcha_query = '''SELECT * FROM scooters WHERE companyname ilike 'gotcha';'''
#jump_query = '''SELECT * FROM scooters WHERE companyname ilike 'jump';'''
#lime_query = '''SELECT * FROM scooters WHERE companyname ilike 'lime' AND sumdgroup != 'bicycle';'''
#lyft_query = '''SELECT * FROM scooters WHERE companyname ilike 'lyft';'''
#spin_query = '''SELECT * FROM scooters WHERE companyname ilike 'spin';'''

# with engine.connect() as connection:
#    bird = pd.read_sql(text(bird_query), con = connection)
#     bolt = pd.read_sql(text(bolt_query), con = connection)
#     gotcha = pd.read_sql(text(gotcha_query), con = connection) 
#     jump = pd.read_sql(text(jump_query), con = connection)
#    lime = pd.read_sql(text(lime_query), con = connection)
#     lyft = pd.read_sql(text(lyft_query), con = connection)
#     spin = pd.read_sql(text(spin_query), con = connection)
    
#bird.to_csv('../data/scooters_bird.csv',index=False)
#bolt.to_csv('../data/scooters_bolt.csv',index=False)
#gotcha.to_csv('../data/scooters_gotcha.csv',index=False)
#jump.to_csv('../data/scooters_jump.csv',index=False)
#lime.to_csv('../data/scooters_lime.csv',index=False)
#lyft.to_csv('../data/scooters_lyft.csv',index=False)
#spin.to_csv('../data/scooters_spin.csv',index=False)


In [7]:
# parse trips data to csv.  Uncomment lines below to write file to project directory; recomment when complete to avoid running query multiple times.

# trips_query = '''SELECT * FROM trips WHERE sumdid NOT IN ('Standard5JXOV277MCWID', 'StandardNPOOZNUSGAXZN', 'StandardNUTLLXP4G37OI','StandardNW5HJFO4R32LY','StandardZPUQESHVPP74J');'''
# with engine.connect() as connection:
#      trips = pd.read_sql(text(trips_query), con = connection)
# trips.to_csv('../data/trips.csv',index=False)


In [8]:
#call in new csv dataframes

bird = pd.read_csv('../data/scooters_bird.csv')
bolt = pd.read_csv('../data/scooters_bolt.csv')
gotcha = pd.read_csv('../data/scooters_gotcha.csv')
jump = pd.read_csv('../data/scooters_jump.csv')
lyft = pd.read_csv('../data/scooters_lyft.csv')
spin = pd.read_csv('../data/scooters_spin.csv')
trips = pd.read_csv('../data/trips.csv')

### EXPLORATION

#### EQ1: Are there any null values in any columns in either table?

In [9]:
# Use "company.info(verbose=True, show_counts=True)" to investicage shapes of parsed data

print("---BIRD INFO---")
print(bird.info(verbose=True, show_counts=True))
print("---BOLT INFO---")
print(bolt.info(verbose=True, show_counts=True))
print("---GOTCHA INFO---")
print(gotcha.info(verbose=True, show_counts=True))
print("---JUMP INFO---")
print(jump.info(verbose=True, show_counts=True))
print("---LYFT INFO---")
print(lyft.info(verbose=True, show_counts=True))
print("---SPIN INFO---")
print(spin.info(verbose=True, show_counts=True))

## charge level: Bolt, Spin

---BIRD INFO---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12251590 entries, 0 to 12251589
Data columns (total 9 columns):
 #   Column       Non-Null Count     Dtype  
---  ------       --------------     -----  
 0   pubdatetime  12251590 non-null  object 
 1   latitude     12251590 non-null  float64
 2   longitude    12251590 non-null  float64
 3   sumdid       12251590 non-null  object 
 4   sumdtype     12251590 non-null  object 
 5   chargelevel  12251590 non-null  float64
 6   sumdgroup    12251590 non-null  object 
 7   costpermin   12251590 non-null  float64
 8   companyname  12251590 non-null  object 
dtypes: float64(4), object(5)
memory usage: 841.3+ MB
None
---BOLT INFO---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3477198 entries, 0 to 3477197
Data columns (total 9 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   pubdatetime  3477198 non-null  object 
 1   latitude     3477198 non-null  float64
 2   longitu

#### EA1. Bolt and Spin have missing some missing values in the chargelevel column

#### EQ2. What date range is represented in each of the date columns? Investigate any values that seem odd.

In [10]:
# What date range is represented in each of the date columns? Investigate any values that seem odd.

#scooters - pubdatetime
scooters_date_query = '''SELECT pubdatetime FROM scooters'''
#trips - pubtimestamp, startdate, starttime, enddate, endtime, create_dt
trips_date_query = '''SELECT pubtimestamp, startdate, starttime, enddate, endtime, create_dt FROM trips'''

with engine.connect() as connection:
    scoot_dates = pd.read_sql(text(scooters_date_query), con = connection)
    trip_dates = pd.read_sql(text(trips_date_query), con = connection)

In [11]:
#scooter dates
#earliest
scootfirstdate = scoot_dates.min()
#latest
scootlastdate = scoot_dates.max()
#range
scootrange = scootlastdate - scootfirstdate

#trips dates
#earliest
trippubfirst = trip_dates.pubtimestamp.min()
tripstartfirst = trip_dates.startdate.min()
tripendfirst = trip_dates.enddate.min()
tripcratefirst = trip_dates.create_dt.min()
#latest dates
trippublast = trip_dates.pubtimestamp.max()
tripstartlast  = trip_dates.startdate.max()
tripendlast  = trip_dates.enddate.max()
tripcratelast  = trip_dates.create_dt.max()

#date ranges
trippubrange = trippublast - trippubfirst
tripstartrange  = tripstartlast - tripstartfirst
tripendrange  = tripendlast - tripendfirst
tripcraterange  = tripcratelast - tripcratefirst

In [12]:
print('scooter table date ranges')
print('scooter - pubdatetime - earliest: ' + str(scootfirstdate))
#latest
print('scooter - pubdatetime - latest: ' + str(scootlastdate))
#range
print('scooter - pubdatetime - range: ' + str(scootrange))

print('trips table date ranges')
#earliest
print('trips - pubdatetime - earliest: ' + str(trippubfirst))
print('trips - starttime - earliest: ' + str(tripstartfirst))
print('trips - endtime - earliest: ' + str(tripendfirst))
print('trips - create time - earliest: ' + str(tripcratefirst))

#latest dates
print('trips - pubdatetime - latest: ' + str(trippublast))
print('trips - starttime - latest: ' + str(tripstartlast))
print('trips - endtime - latest: ' + str(tripendlast))
print('trips - create time - latest: ' + str(tripcratelast))

#date ranges
print('trips - pubdatetime - range: ' + str(trippubrange))
print('trips - starttime - range: ' + str(tripstartrange))
print('trips - endtime - range: ' + str(tripendrange))
print('trips - create time - range: ' + str(tripcraterange))

scooter table date ranges
scooter - pubdatetime - earliest: pubdatetime   2019-05-01 00:01:41.247
dtype: datetime64[ns]
scooter - pubdatetime - latest: pubdatetime   2019-07-31 23:59:57
dtype: datetime64[ns]
scooter - pubdatetime - range: pubdatetime   91 days 23:58:15.753000
dtype: timedelta64[ns]
trips table date ranges
trips - pubdatetime - earliest: 2019-05-01 00:00:55.423000
trips - starttime - earliest: 2019-05-01
trips - endtime - earliest: 2019-05-01
trips - create time - earliest: 2019-05-02 05:30:23.780000
trips - pubdatetime - latest: 2019-08-01 07:04:00
trips - starttime - latest: 2019-07-31
trips - endtime - latest: 2019-08-01
trips - create time - latest: 2019-08-02 11:30:29.923000
trips - pubdatetime - range: 92 days 07:03:04.577000
trips - starttime - range: 91 days, 0:00:00
trips - endtime - range: 92 days, 0:00:00
trips - create time - range: 92 days 06:00:06.143000


#### EA2.
Some fields are date only and format varies. Date range is from May 1, 2019 to July 31, 2019.

#### EQ3. Is time represented with am/pm or using 24 hour values in each of the columns that include time?


In [13]:
#investigate max dates
scoot_dates.pubdatetime.max()

Timestamp('2019-07-31 23:59:57')

In [14]:
trip_dates.starttime.max()

datetime.time(23, 59, 59, 506666)

#### EA3.
Time is represented in 24 hour values: I%:M%:S%:f%

#### EQ4. What values are there in the sumdgroup column? Are there any that are not of interest for this project?

In [15]:
# What values are there in the sumdgroup column? Are there any that are not of interest for this project?
sumd_query = '''SELECT DISTINCT sumdgroup FROM scooters'''
with engine.connect() as connection:
    sumd_result = pd.read_sql(text(sumd_query), con = connection)
sumd_result

sumdgroup
0   bicycle
1   scooter
2   Scooter

In [16]:
# find bicycles 

bicycle_query = '''SELECT DISTINCT sumdid, companyname FROM scooters WHERE sumdgroup = 'bicycle';'''
with engine.connect() as connection:
    bikes = pd.read_sql(text(bicycle_query), con = connection)
bikes

sumdid companyname
0  Standard2UGJKREVB53HT        Lime
1  Standard5JXOV277MCWID        Lime
2  StandardNPOOZNUSGAXZN        Lime
3  StandardNUTLLXP4G37OI        Lime
4  StandardNW5HJFO4R32LY        Lime
5  StandardZPUQESHVPP74J        Lime

In [17]:
print(bikes.count())

sumdid         6
companyname    6
dtype: int64


#### EA4. The dataset contiains entires for six Lime bicycles; these entries should be omitted from scooter analysis

#### EQ5. What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [18]:
# What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising? -What is the range of values for trip duration and trip distance? Do these values make sense? Explore values that might seem questionable.

sll_query = '''SELECT
                MIN(latitude) as min_latitude,
                MAX(latitude) as max_latitude,
                MIN(longitude) as min_longitude,
                MAX(latitude) as max_longitude
            FROM scooters'''
            
tll_query = '''SELECT
                MIN(startlatitude) as min_latitude_start,
                MAX(startlatitude) as max_latitude_start,
                MIN(startlongitude) as min_longitude_start,
                MAX(startlongitude) as max_longitude_start,
                MIN(endlatitude) as min_latitude_end,
                MAX(endlatitude) as max_latitude_end,
                MIN(endlongitude) as min_longitude_end,
                MAX(endlongitude) as max_longitude_end
            FROM trips'''

with engine.connect() as connection:
    scoot_result = pd.read_sql(text(sll_query), con = connection)
    trip_result = pd.read_sql(text(tll_query), con = connection)
    


In [19]:
scoot_result

min_latitude  max_latitude  min_longitude  max_longitude
0           0.0  3.609874e+06     -97.443879   3.609874e+06

In [20]:
trip_result

min_latitude_start  max_latitude_start  min_longitude_start  \
0             35.8532           36.300029           -86.918008   

   max_longitude_start  min_latitude_end  max_latitude_end  min_longitude_end  \
0             -86.3662        -36.850405         51.045409        -122.673729   

   max_longitude_end  
0         174.764886

#### EA5.  The scooters table has a minimum latitude value of 0.0.  Latitude and longitude are expressed in defferent formats in the scooters and trips tables.

#### EQ6. Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?

In [21]:
# Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?
scooter_companies = '''SELECT DISTINCT companyname FROM scooters'''
trips_companies = ''' SELECT DISTINCT companyname FROM trips'''

with engine.connect() as connection:
    scompanies = pd.read_sql(text(scooter_companies), con = connection)
    tcompanies = pd.read_sql(text(trips_companies), con = connection)

In [22]:
scompanies.companyname.tolist()

['Bird', 'Bolt', 'Gotcha', 'Jump', 'Lime', 'Lyft', 'Spin']

In [23]:
tcompanies.companyname.tolist()

['Bird', 'Bolt Mobility', 'Gotcha', 'JUMP', 'Lime', 'Lyft', 'SPIN']

#### EA6. The same companies are present in the scooters and trips tables, however there are some variations in capitalization.  Scooters contains company name "Bolt"; trips contains company name "Bolt Mobility"

Scooters: ['Bird', 'Bolt', 'Gotcha', 'Jump', 'Lime', 'Lyft', 'Spin']

Trips: ['Bird', 'Bolt Mobility', 'Gotcha', 'JUMP', 'Lime', 'Lyft', 'SPIN']

## ANALYSIS

#### Questions for consideration:
1. During this period, seven companies offered scooters. How many scooters did each company have in this time frame? Did the number for each company change over time? Did scooter usage vary by company?
<br>
<br>
2. According to Second Substitute Bill BL2018-1202 (as amended) (https://web.archive.org/web/20181019234657/https://www.nashville.gov/Metro-Clerk/Legislative/Ordinances/Details/7d2cf076-b12c-4645-a118-b530577c5ee8/2015-2019/BL2018-1202.aspx), all permitted operators will first clean data before providing or reporting data to Metro. Data processing and cleaning shall include:<br>
-- Removal of staff servicing and test trips<br>
-- Removal of trips below one minute<br>
-- Trip lengths are capped at 24 hours<br>
Are the scooter companies in compliance with the second and third part of this rule?
<br>
<br>
3. The goal of Metro Nashville is to have each scooter used a minimum of 3 times per day. Based on the data, what is the average number of trips per scooter per day? Make sure to consider the days that a scooter was available. How does this vary by company?
<br>
<br>
4. Metro would like to know how many scooters are needed, and something that could help with this is knowing peak demand. Estimate the highest count of scooters being used at the same time. When were the highest volume times? Does this vary by zip code or other geographic region?
<br>
<br>
5. Stretch Goal: SUMDs can provide alternative transportation and provide "last mile" access to public transit. How often are trips starting near public transit hubs? You can download a dataset of bus stop locations from https://data.nashville.gov/Transportation/WeGo-Transit-Bus-Stops/vfe9-k7vc/about_data.

#### Q1:

In [24]:
#How many scooters did each company have
query = '''SELECT companyname, count(DISTINCT(sumdid))
    FROM scooters
    WHERE sumdgroup NOT LIKE 'bicycle'
    GROUP BY companyname'''

with engine.connect() as connection:
    company_count = pd.read_sql(text(query), con = connection)
    
company_count.sort_values('count', ascending=False)

companyname  count
0        Bird   3860
4        Lime   1824
5        Lyft   1735
3        Jump   1210
6        Spin    805
1        Bolt    360
2      Gotcha    224

In [ ]:
#Did the number for each company change over time?
query = '''SELECT companyname, COUNT(DISTINCT(sumdid)), EXTRACT(Month from pubdatetime) as month, EXTRACT(Day from pubdatetime) as day
            FROM scooters
            WHERE sumdgroup NOT LIKE 'bicycle'
            GROUP BY companyname, month, day;'''
with engine.connect() as connection:
    scooters_over_time = pd.read_sql(text(query), con = connection)
    
#scooters_over_time.date = pd.to_datetime(scooters_over_time.date, format = 'Y%:%m:%d')
    
scooters_over_time.sort_values(['month', 'day'], ascending=[True,True])

In [ ]:
#Did scooter usage vary by company?
#

#### Q2:

In [71]:
trips.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance  startdate                  starttime  \
0        3.0000     958.00528 2019-05-01 2024-04-25 00:00:20.460000   
1        1.7156    1371.39112 2019-05-01 2024-04-25 00:01:50.090000   
2        3.0000    2296.58800 2019-05-01 2024-04-25 00:03:47.363333   
3        3.0000    1200.78744 2019-05-01 2024-04-25 00:04:21.386666   
4        2.0000     351.04988 2019-05-01 2024-04-25 00:04:27.796666   

     enddate                    endtime  startlatitude  startlongitude  \
0 2019-05-01 2024-04-25 00:02:52.346666       36.15710       -86.80360   
1 2019-05-01 2024-04-25 00:03:33.026666       36.15797       -86.77896   
2 2019-05-01 2024-04-25 00:07:13.596666       36.15470       -86.78180   
3 2019-05-01 2024-04-25 00:06:59.176666       36.14940       -86.77950   
4 2019-05-01 2024-04-25 00:06:23.150000       36.17780       -86.78660   

   endlatitude  endlongitude  \
0     36.15660     -86.80670   
1     36.16054     -86.77689   
2     36.15650     -86.78680   
3     36.15310     -86.77960   
4     36.17740     -86.78760   

                                           triproute                create_dt  \
0  [(36.157235, -86.803612), (36.157235, -86.8036...  2019-05-02 05:30:23.780   
1  [(36.15797, -86.77896), (36.15795, -86.77873),...  2019-05-02 07:20:32.757   
2  [(36.155068, -86.782124), (36.156597, -86.78675)]  2019-05-02 05:30:24.530   
3  [(36.149741, -86.779344), (36.149741, -86.7793...  2019-05-02 05:30:24.237   
4  [(36.177699, -86.786477), (36.177711, -86.7864...  2019-05-02 05:30:24.207   

            tripstart             tripend   duration_full  trip_minutes  
0 2019-05-01 00:00:20 2019-05-01 00:02:52 0 days 00:02:32      2.533333  
1 2019-05-01 00:01:50 2019-05-01 00:03:33 0 days 00:01:43      1.716667  
2 2019-05-01 00:03:47 2019-05-01 00:07:13 0 days 00:03:26      3.433333  
3 2019-05-01 00:04:21 2019-05-01 00:06:59 0 days 00:02:38      2.633333  
4 2019-05-01 00:04:27 2019-05-01 00:06:23 0 days 00:01:56      1.933333

In [72]:
print(trips.tripduration.min())
print(trips.tripduration.max())


-19.3582666667
512619.0


In [73]:
sub1 = trips[trips.tripduration < 1]
sub1.shape

(9154, 20)

In [74]:
over24 = trips[trips.tripduration > 1440]
over24.shape

(6938, 20)

In [75]:
trips.dtypes

pubtimestamp       datetime64[ns]
companyname                object
triprecordnum              object
sumdid                     object
tripduration              float64
tripdistance              float64
startdate          datetime64[ns]
starttime          datetime64[ns]
enddate            datetime64[ns]
endtime            datetime64[ns]
startlatitude             float64
startlongitude            float64
endlatitude               float64
endlongitude              float64
triproute                  object
create_dt                  object
tripstart          datetime64[ns]
tripend            datetime64[ns]
duration_full     timedelta64[ns]
trip_minutes              float64
dtype: object

In [76]:
#transform data

trips.pubtimestamp = pd.to_datetime(trips.pubtimestamp)
trips.startdate = pd.to_datetime(trips.startdate)
trips.starttime = pd.to_datetime(trips.starttime, format = '%I:%M:%S.%f')
trips.enddate = pd.to_datetime(trips.enddate)
trips.endtime = pd.to_datetime(trips.endtime, format = '%I:%M:%S.%f')

#combine datetimes to new columns and convert to datetime objects

trips['tripstart'] = trips['startdate'].dt.strftime('%Y-%m-%d') + ' ' + trips['starttime'].dt.strftime('%H:%M:%S')
trips['tripend'] = trips['enddate'].dt.strftime('%Y-%m-%d') + ' ' + trips['endtime'].dt.strftime('%H:%M:%S')

trips.tripstart = pd.to_datetime(trips.tripstart)
trips.tripend = pd.to_datetime(trips.tripend)

print(trips.dtypes)

pubtimestamp       datetime64[ns]
companyname                object
triprecordnum              object
sumdid                     object
tripduration              float64
tripdistance              float64
startdate          datetime64[ns]
starttime          datetime64[ns]
enddate            datetime64[ns]
endtime            datetime64[ns]
startlatitude             float64
startlongitude            float64
endlatitude               float64
endlongitude              float64
triproute                  object
create_dt                  object
tripstart          datetime64[ns]
tripend            datetime64[ns]
duration_full     timedelta64[ns]
trip_minutes              float64
dtype: object


In [77]:
trips.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance  startdate                  starttime  \
0        3.0000     958.00528 2019-05-01 2024-04-25 00:00:20.460000   
1        1.7156    1371.39112 2019-05-01 2024-04-25 00:01:50.090000   
2        3.0000    2296.58800 2019-05-01 2024-04-25 00:03:47.363333   
3        3.0000    1200.78744 2019-05-01 2024-04-25 00:04:21.386666   
4        2.0000     351.04988 2019-05-01 2024-04-25 00:04:27.796666   

     enddate                    endtime  startlatitude  startlongitude  \
0 2019-05-01 2024-04-25 00:02:52.346666       36.15710       -86.80360   
1 2019-05-01 2024-04-25 00:03:33.026666       36.15797       -86.77896   
2 2019-05-01 2024-04-25 00:07:13.596666       36.15470       -86.78180   
3 2019-05-01 2024-04-25 00:06:59.176666       36.14940       -86.77950   
4 2019-05-01 2024-04-25 00:06:23.150000       36.17780       -86.78660   

   endlatitude  endlongitude  \
0     36.15660     -86.80670   
1     36.16054     -86.77689   
2     36.15650     -86.78680   
3     36.15310     -86.77960   
4     36.17740     -86.78760   

                                           triproute                create_dt  \
0  [(36.157235, -86.803612), (36.157235, -86.8036...  2019-05-02 05:30:23.780   
1  [(36.15797, -86.77896), (36.15795, -86.77873),...  2019-05-02 07:20:32.757   
2  [(36.155068, -86.782124), (36.156597, -86.78675)]  2019-05-02 05:30:24.530   
3  [(36.149741, -86.779344), (36.149741, -86.7793...  2019-05-02 05:30:24.237   
4  [(36.177699, -86.786477), (36.177711, -86.7864...  2019-05-02 05:30:24.207   

            tripstart             tripend   duration_full  trip_minutes  
0 2019-05-01 00:00:20 2019-05-01 00:02:52 0 days 00:02:32      2.533333  
1 2019-05-01 00:01:50 2019-05-01 00:03:33 0 days 00:01:43      1.716667  
2 2019-05-01 00:03:47 2019-05-01 00:07:13 0 days 00:03:26      3.433333  
3 2019-05-01 00:04:21 2019-05-01 00:06:59 0 days 00:02:38      2.633333  
4 2019-05-01 00:04:27 2019-05-01 00:06:23 0 days 00:01:56      1.933333

In [35]:
#calculate trip durations
trips['duration_full'] = trips['tripend'] - trips['tripstart']
trips.dtypes
trips['trip_minutes'] = trips.duration_full / dt.timedelta(minutes=1)

In [78]:
#define compliance variables

#Trips Less Than One Minute
trips_sub1min = trips[trips['trip_minutes'] < 1]
print(trips_sub1min.info())

#Trips Greater than 24 Hours
trips_over24h = trips[trips['trip_minutes'] > 1440]
print(trips_over24h.info())

#Noncompliant Trips
trips_noncompliant = trips_sub1min._append(trips_over24h, ignore_index=True)

#noncompliant companies
noncompliant_companies = trips_noncompliant['companyname'].unique()



<class 'pandas.core.frame.DataFrame'>
Index: 12384 entries, 15 to 565405
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype          
---  ------          --------------  -----          
 0   pubtimestamp    12384 non-null  datetime64[ns] 
 1   companyname     12384 non-null  object         
 2   triprecordnum   12384 non-null  object         
 3   sumdid          12384 non-null  object         
 4   tripduration    12384 non-null  float64        
 5   tripdistance    12384 non-null  float64        
 6   startdate       12384 non-null  datetime64[ns] 
 7   starttime       12384 non-null  datetime64[ns] 
 8   enddate         12384 non-null  datetime64[ns] 
 9   endtime         12384 non-null  datetime64[ns] 
 10  startlatitude   12384 non-null  float64        
 11  startlongitude  12384 non-null  float64        
 12  endlatitude     12384 non-null  float64        
 13  endlongitude    12384 non-null  float64        
 14  triproute       12384 non-null  object   

In [79]:
# DURATION ANALYSIS BY COMPNAY SUB MIN
(trips_sub1min
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
 .reset_index()
 )

companyname  triprecordnum
0           Bird           6377
1  Bolt Mobility            270
2           Lime           1250
3           Lyft           4487

In [80]:
# DURATION ANALYSIS OVER 24 H
(trips_over24h
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
 .reset_index()
 )

companyname  triprecordnum
0  Bolt Mobility              4
1           JUMP             27
2           Lyft              2
3           SPIN             28

In [40]:
# DURATION ANALYSIS ALL NONCOMPLIANT
(trips_noncompliant
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
  .reset_index()
 .sort_values('triprecordnum', ascending=False)
 )

companyname  triprecordnum
0           Bird           6377
4           Lyft           4489
3           Lime           1250
1  Bolt Mobility            274
5           SPIN             28
2           JUMP             27

In [81]:
#totals

totaltrips = len(trips.index)
totalsub1min = len(trips_sub1min.index)
totalover24h = len(trips_over24h.index)
totalnoncompliant = totalsub1min + totalover24h
percentagenoncompliant = totalnoncompliant/totaltrips * 100

print('Total number of scooter trips reported: ' + str(totaltrips))
print('Total number of scooter trips with a duration less than one minute: ' + str(totalsub1min))
print('Total number of scooter trips with a duration over 24 hours: ' + str(totalover24h))

print('Total number of noncompliant scooter trips: ' + str(totalnoncompliant))
print('Percentage of noncompliant scooter trips: ' + str(percentagenoncompliant))

print('Companies with noncompliant scooter trip data: ' + str(noncompliant_companies))

Total number of scooter trips reported: 565522
Total number of scooter trips with a duration less than one minute: 12384
Total number of scooter trips with a duration over 24 hours: 61
Total number of noncompliant scooter trips: 12445
Percentage of noncompliant scooter trips: 2.2006217264757164
Companies with noncompliant scooter trip data: ['Bird' 'Lyft' 'Lime' 'Bolt Mobility' 'SPIN' 'JUMP']


#### A2: 
Total number of scooter trips reported: 565522 <br>
Total number of scooter trips with a duration less than one minute: 12384 <br>
Total number of scooter trips with a duration over 24 hours: 61 <br>
Total number of noncompliant scooter trips: 12445 <br>
Percentage of noncompliant scooter trips: 2.2006217264757164 <br>
Companies with noncompliant scooter trip data: ['Bird' 'Lyft' 'Lime' 'Bolt Mobility' 'SPIN' 'JUMP'] <br>

In [88]:
#pre update check
trips.shape

(565522, 20)

In [91]:
#Update the trips dataframe to remove any trips under one minute
trips = trips.drop(trips[trips.trip_minutes < 1].index)
#post update check
trips.shape

In [94]:
#Update the trips dataframe to remove any trips over 1440 minutes (24 hours = 1440 minutes)
trips = trips.drop(trips[trips.trip_minutes > 1440].index)
#post update check
trips.shape

(553077, 20)

#### Q3: 

In [ ]:
# The goal of Metro Nashville is to have each scooter used a minimum of 3 times per day. 
# Based on the data, what is the average number of trips per scooter per day? 

#scooterid -- scoooters: DISTINCT(sumdid)
#total number of trips -- trips: count triprecordnum by sumdid
#total number of days active -- scooters: count of days (YYYY MM DD) by scooter id
#average trips per day -- count triprecordnum by sumdid / count of days (YYYY MM DD) by scooter id AS avg_trips


query = '''
SELECT DISTICNT
    s.sumdid, 
    COUNT(t.triprecordnum) as number_of_trips
    COUNT(t.tripstart) as tripcount,
    COUNT(s.) as daysactive,
    t.COUNT(triprecordnum)/s.COUNT(emilysmonthdaycolumn) as avgtripsperday
  
FROM scooters as s

LEFT JOIN trips s
on t.sumdid = s.sumdid;'''

# query = '''
# SELECT DISTICNT
#     t.sumdid as unit, 
#     t.companyname
#     t.COUNT(triprecordnum) as tripcount,
#     s.COUNT(emilysmonthdaycolumn) as daysactive,
#     t.COUNT(triprecordnum)/s.COUNT(emilysmonthdaycolumn) as avgtripsperday
  
# FROM trips as t

# LEFT JOIN scooters s
# on t.sumdid = s.sumdid;'''

In [ ]:
# Make sure to consider the days that a scooter was available. How does this vary by company?
groupby statement in python

#### Q4:

In [ ]:
# Metro would like to know how many scooters are needed, and something that could help with this is knowing peak demand. 
# Estimate the highest count of scooters being used at the same time. 
# When were the highest volume times? Does this vary by zip code or other geographic region?

